tensorboard/notebook.py

```diff
@@ -378,18 +379,35 @@ def _display_ipython(port, height, display_handle):
       <script>
         (function() {
           const frame = document.getElementById(%JSON_ID%);
-          const url = new URL("/", window.location);
-          url.port = %PORT%;
+          const url = new URL(%URL%, window.location);
+          const port = %PORT%;
+          if (port) {
+            url.port = port;
+          }
           frame.src = url;
         })();
       </script>
-  """
-    replacements = [
-        ("%HTML_ID%", html_escape(frame_id, quote=True)),
-        ("%JSON_ID%", json.dumps(frame_id)),
-        ("%PORT%", "%d" % port),
-        ("%HEIGHT%", "%d" % height),
-    ]
+    """
+    proxy_url = os.environ.get("TENSORBOARD_PROXY_URL")
+    if proxy_url is not None:
+        # Allow %PORT% in $TENSORBOARD_PROXY_URL
+        proxy_url = proxy_url.replace("%PORT%", "%d" % port)
+        replacements = [
+            ("%HTML_ID%", html_escape(frame_id, quote=True)),
+            ("%JSON_ID%", json.dumps(frame_id)),
+            ("%URL%", json.dumps(proxy_url)),
```

Comment:
Could you move `%URL%` to the end of the substitutions list just to
avoid any issues in case the environment variable happens to contain the
strings `%PORT%` or `%HEIGHT%`? (likewise below since this substitution
table appears twice)